### PR TITLE
Mistral configs

### PR DESCRIPTION
**Pull request by kartikayk**
## Context

#571 added support for Mistral 7B. In this PR, I add configs for Mistral7B full finetuning and lora. These can be further improved, but have decent results OOTB.

Full finetune:
```
 tune --nnodes 1 --nproc_per_node 4 full_finetune_distributed \
--config mistral/7B_full \
metric_logger=torchtune.utils.metric_logging.WandBLogger \
metric_logger.project=test
```

Loss Curve:

Screenshot 2024-03-25 at 9 00 57 AM

Eval on ```truthfulqa_mc2```

Screenshot 2024-03-25 at 9 01 08 AM

LoRA: 
```
tune --nnodes 1 --nproc_per_node 4 lora_finetune_distributed \
--config mistral/7B_lora  \
metric_logger=torchtune.utils.metric_logging.WandBLogger \
metric_logger.project=test
```

Loss Curve:

image

Eval on ```truthfulqa_mc2```

Screenshot 2024-03-26 at 7 11 15 AM

## FAQ

How did you come up with these configs?

Getting training to stabilize took some work. This model seems to need a smaller LR than the Llama2 7B/13B models and for LoRA, I needed to ramp up rank and alpha a bit. I don't claim this to be novel work though. I just did a bunch of snooping around on localllama [[example](https://www.reddit.com/r/LocalLLaMA/comments/1amjzjw/mistral_finetuning_loss_increases_crazy/)] and some other forums and blogs to come up with a config with reasonable results.

## Changelog
- Adding component and model builders for LoRA with Mistral 7B
- Add configs for LoRA and Full finetuning for Mistral 7B

## Test plan
- Tests

```
pytest tests
```

- E2E Runs - see above

### CODE REVIEW

File: torchtune/models/mistral/_component_builders.py [hunk 1]
```diff
@@ -115,3 +117,254 @@ def mistral_mlp(dim: int, hidden_dim: int) -> FeedForward:
     down_proj = nn.Linear(hidden_dim, dim, bias=False)
     up_proj = nn.Linear(dim, hidden_dim, bias=False)
     return FeedForward(gate_proj=gate_proj, down_proj=down_proj, up_proj=up_proj)
+
+
+def lora_mistral(
+    lora_attn_modules: List[LORA_ATTN_MODULES],
+    apply_lora_to_mlp: bool = False,
+    apply_lora_to_output: bool = False,
+    *,
```
Comment (joecummings) [inline torchtune/models/mistral/_component_builders.py:126]: wooooo kwargs

File: recipes/configs/mistral/7B_lora.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# This config is currently a WIP. Use it with caution
```
Comment (joecummings) [inline recipes/configs/mistral/7B_lora.yaml:1]: Are these configs based on the paper and/or other results showing good performance?

Comment (kartikayk) [reply recipes/configs/mistral/7B_lora.yaml:1]: Mostly digging on various forums like reddit

File: recipes/configs/mistral/7B_full.yaml [hunk 3]
```diff
@@ -25,16 +25,16 @@ checkpointer:
     pytorch_model-00002-of-00002.bin
   ]
   recipe_checkpoint: null
-  output_dir: /tmp/Mistral-7B-v0.1
-  model_type: LLAMA2
+  output_dir: /tmp/Mistral-7B-v0.1/
+  model_type: MISTRAL
```
Comment (joecummings) [inline recipes/configs/mistral/7B_full.yaml:29]: This technically varies from the HF way of defining model type b/c mistral is llama-based. What does model_type=Mistral give us?

Comment (kartikayk) [reply recipes/configs/mistral/7B_full.yaml:29]: Less confusion? Defining mistral model to be llama type is a bit confusing I think

Comment (joecummings) [reply recipes/configs/mistral/7B_full.yaml:29]: But it's technically true that it is a llama-type model and this is the accepted standard in HF.

Comment (kartikayk) [reply recipes/configs/mistral/7B_full.yaml:29]: I'd definitely dont want to carry over any confusions from there

File: README.md [hunk 4]
```diff
@@ -40,7 +40,7 @@ The library currently supports the following models and fine-tuning methods.
 |-----------------------------------------------|-----------|-----------------------------------------------------------|
 | [Llama2](torchtune/models/llama2/_model_builders.py)   | 7B        | Full Finetuning [[single device](recipes/configs/llama2/7B_full_single_device.yaml),  [distributed](recipes/configs/llama2/7B_full.yaml)] LoRA [[single device](recipes/configs/llama2/7B_lora_single_device.yaml),  [distributed](recipes/configs/llama2/7B_lora.yaml)] QLoRA [single device](recipes/configs/llama2/7B_qlora_single_device.yaml) |
 | [Llama2](torchtune/models/llama2/_model_builders.py)   | 13B       | [Full Finetuning](recipes/configs/llama2/13B_full.yaml), [LoRA](recipes/configs/llama2/13B_lora.yaml)
-| [Mistral](torchtune/models/mistral//_model_builders.py)   | 7B       | Full Finetuning and LoRA are WIP and will be added soon
+| [Mistral](torchtune/models/mistral//_model_builders.py)   | 7B       | [Full Finetuning](recipes/configs/mistral/7B_full.yaml), [LoRA](recipes/configs/mistral/7B_lora.yaml)
```
Comment (ebsmothers) [inline README.md:43]: Not necessary for this PR but I think this table needs a cleanup. We should either split finetuning methods into separate columns or rows grouped under model family

File: torchtune/models/mistral/_component_builders.py [hunk 5]
```diff
@@ -115,3 +117,254 @@ def mistral_mlp(dim: int, hidden_dim: int) -> FeedForward:
...
+    layer = TransformerDecoderLayer(
+        attn=self_attn,
+        mlp=mlp,
+        sa_norm=RMSNorm(dim=embed_dim, eps=norm_eps),
+        mlp_norm=RMSNorm(dim=embed_dim, eps=norm_eps),
+    )
+
+    tok_embeddings = nn.Embedding(vocab_size, embed_dim)
+
+    # TODO: quantize_base is not applied to final output_proj currently.
```
Comment (rohan-varma) [inline torchtune/models/mistral/_component_builders.py:211]: We don't have quantize_base at all in mistral right now, right?

File: torchtune/models/mistral/_model_builders.py [hunk 6]
```diff
@@ -46,3 +47,52 @@ def mistral_tokenizer(path: str) -> Tokenizer:
...
+
+def lora_mistral_7b(
+    lora_attn_modules: List[LORA_ATTN_MODULES],
+    apply_lora_to_mlp: bool = False,
+    apply_lora_to_output: bool = False,
+    lora_rank: int = 8,
+    lora_alpha: float = 16,
+) -> TransformerDecoder:
+    """
+    Builder for creating a Llama2 7B model with LoRA enabled.
```
Comment (rohan-varma) [inline torchtune/models/mistral/_model_builders.py:60]: :D

File: torchtune/models/mistral/_component_builders.py (hunk 5 above)
Comment (kartikayk) [reply torchtune/models/mistral/_component_builders.py:211]: Yeh need to run some experiments on this before I add this. Training on mistral is quite different from llama2 and so will do QLoRA as a follow up

File: recipes/configs/mistral/7B_full.yaml (hunk 3 above)
Comment (ebsmothers) [reply recipes/configs/mistral/7B_full.yaml:29]: Personally I like the separate model type. Yes it's an identical architecture, but just helps to be explicit. Main q is: is this used as a hard constraint in places? E.g. is it gonna prevent me from loading Mistral weights into a LLAMA2 model type

File: torchtune/models/mistral/_component_builders.py [hunk 7]
```diff
@@ -115,3 +117,254 @@ def mistral_mlp(dim: int, hidden_dim: int) -> FeedForward:
...
+def lora_mistral_self_attention(
+    lora_modules: List[LORA_ATTN_MODULES],
+    *,
+    # CausalSelfAttention args
+    embed_dim: int,
+    num_heads: int,
+    num_kv_heads: int,
+    max_seq_len: int,
+    attn_dropout: float = 0.0,
+    rope_base: int = 10_000,
```
Comment (ebsmothers) [inline torchtune/models/mistral/_component_builders.py:237]: Why do we parametrize this here but not in llama2 builders?

File: torchtune/models/mistral/_model_builders.py [hunk 8]
```diff
@@ -46,3 +47,52 @@ def mistral_tokenizer(path: str) -> Tokenizer:
...
+        apply_lora_to_output (bool): whether to apply LoRA to the model's final output projection.
+            Default: False
+        lora_rank (int): rank of each low-rank approximation
+        lora_alpha (float): scaling factor for the low-rank approximation
+        max_batch_size (Optional[int]): Maximum batch size to be passed to KVCache.
+
+    Returns:
+        TransformerDecoder: Instantiation of Llama2 7B model with LoRA applied
+    """
+    return lora_mistral(
```
Comment (ebsmothers) [inline torchtune/models/mistral/_model_builders.py:81]: Would be nice to document somewhere exactly what the differences are between this builder and the equivalent llama2 one. Doesn't have to be for this PR though; as an added bonus it nicely shows off how easily we can switch between the two.

File: recipes/configs/mistral/7B_lora.yaml [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+dataset:
+  _component_: torchtune.datasets.alpaca_dataset
+  train_on_input: True
+seed: null
+shuffle: True
+
+# Model Arguments
+model:
+  _component_: torchtune.models.mistral.lora_mistral_7b
+  lora_attn_modules: ['q_proj', 'v_proj', 'k_proj']
```
Comment (ebsmothers) [inline recipes/configs/mistral/7B_lora.yaml:19]: nit: order q, k, v for clarity

File: recipes/configs/mistral/7B_lora.yaml [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  _component_: torch.nn.CrossEntropyLoss
+max_steps_per_epoch: null
+gradient_accumulation_steps: 1
+
+
+# Training env
+device: cuda
+
+# Distributed
+cpu_offload: False
```
Comment (ebsmothers) [inline recipes/configs/mistral/7B_lora.yaml:56]: I don't think this is even used? (I see it's still lurking around in other configs too..)

File: torchtune/models/mistral/_component_builders.py (hunk 5 above)
Comment (ebsmothers) [reply torchtune/models/mistral/_component_builders.py:211]: Would at least remove the todo for now then

**Review by ebsmothers — COMMENTED** [review]
Do we wanna add a unit test for the new model? Doesn't have to be for all levels of components but at least e.g. at the level of `mistral` or `lora_mistral`

File: README.md (hunk 4 above)
Comment (kartikayk) [reply README.md:43]: Agreed, this table does need an update. Let me follow up separately on this.

**Review by ebsmothers — APPROVED** [review]
